consensus/polybft/extra.go

```diff
@@ -0,0 +1,435 @@
+	Validators *ValidatorSetDelta
+	Seal       []byte
+	Parent     *Signature
+	Committed  *Signature
+}
+
+// MarshalRLPTo defines the marshal function wrapper for Extra
+func (i *Extra) MarshalRLPTo(dst []byte) []byte {
+	ar := &fastrlp.Arena{}
+
+	return i.MarshalRLPWith(ar).MarshalTo(dst)
+}
+
+// MarshalRLPWith defines the marshal function implementation for Extra
+func (i *Extra) MarshalRLPWith(ar *fastrlp.Arena) *fastrlp.Value {
+	vv := ar.NewArray()
+
+	// Validators
+	if i.Validators == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Validators.MarshalRLPWith(ar))
+	}
+
+	// Seal
+	if len(i.Seal) == 0 {
+		vv.Set(ar.NewNull())
+	} else {
+		vv.Set(ar.NewBytes(i.Seal))
+	}
+
+	// ParentSeal
+	if i.Parent == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Parent.MarshalRLPWith(ar))
+	}
+
+	// CommittedSeal
+	if i.Committed == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Committed.MarshalRLPWith(ar))
+	}
+
+	return vv
+}
+
+// UnmarshalRLP defines the unmarshal function wrapper for Extra
+func (i *Extra) UnmarshalRLP(input []byte) error {
+	return fastrlp.UnmarshalRLP(input, i)
+}
+
+// UnmarshalRLPWith defines the unmarshal implementation for Extra
+func (i *Extra) UnmarshalRLPWith(v *fastrlp.Value) error {
+	elems, err := v.GetElems()
+	if err != nil {
+		return err
+	}
+
+	if num := len(elems); num != 4 {
+		return fmt.Errorf("not enough elements to decode extra, expected 4 but found %d", num)
+	}
+
+	// Validators
+	{
+		if elems[0].Elems() > 0 {
+			i.Validators = &ValidatorSetDelta{}
+			if err := i.Validators.UnmarshalRLPWith(elems[0]); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Seal
+	{
+		if elems[1].Len() > 0 {
+			if i.Seal, err = elems[1].GetBytes(i.Seal); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Parent
+	{
+		if elems[2].Elems() > 0 {
+			i.Parent = &Signature{}
+			if err := i.Parent.UnmarshalRLPWith(elems[2]); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Committed
+	{
+		if elems[3].Elems() > 0 {
+			i.Committed = &Signature{}
+			if err := i.Committed.UnmarshalRLPWith(elems[3]); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// CreateValidatorSetDelta calculates ValidatorSetDelta based on the provided old and new validator sets
+func CreateValidatorSetDelta(log hclog.Logger, oldValidatorSet, newValidatorSet AccountSet) *ValidatorSetDelta {
+	var addedValidators AccountSet
+
+	oldValidatorSetMap := make(map[types.Address]*ValidatorAccount)
+	removedValidators := map[types.Address]int{}
+
+	for i, validator := range oldValidatorSet {
+		if (validator.Address != types.Address{}) {
+			removedValidators[validator.Address] = i
+			oldValidatorSetMap[validator.Address] = validator
+		}
+	}
+
+	for _, newValidator := range newValidatorSet {
+		// Check if the validator is among both old and new validator set
+		oldValidator, ok := oldValidatorSetMap[newValidator.Address]
+		if ok {
+			if !oldValidator.Equals(newValidator) {
+				log.Error(
```

Comment:
Is it ok to continue after this error?

Reply:
To our understanding, there should never be a use case that existing account with given ECDSA address changes its public BLS key (and that is something which this if is checking). We just wanted to have some indication if that happens for some reason (just in case).

However we are open for discussing the best approach, some of the options are:
1. remove entirely that condition, since it is merely impossible to happen
2. return an error (or maybe even panic)
3. reduce log level to `warning`

WDYT?

Reply:
panic is worst option because someone can trigger that and cause failing of all other nodes. First thing is also not good

Reply:
I agree with @JekaMas, it's useless to have this conditional if it can never happen in the first place - drop it

Reply:
Never say impossible 😛 , can a validator change it's BLS key manually?